internal/datastore/postgres/strictreader.go

```diff
@@ -0,0 +1,65 @@
+package postgres
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+
+	"github.com/authzed/spicedb/internal/datastore/common"
+	pgxcommon "github.com/authzed/spicedb/internal/datastore/postgres/common"
+)
+
+const pgInvalidArgument = "22023"
+
+// strictReaderQueryFuncs wraps a DBFuncQuerier and adds a strict read assertion to all queries.
+// This assertion ensures that the transaction is not reading from the future or from a
+// transaction that has not been committed on the replica.
+type strictReaderQueryFuncs struct {
+	wrapped  pgxcommon.DBFuncQuerier
+	revision postgresRevision
+}
+
+func (srqf strictReaderQueryFuncs) ExecFunc(ctx context.Context, tagFunc func(ctx context.Context, tag pgconn.CommandTag, err error) error, sql string, args ...any) error {
+	// NOTE: it is *required* for the pgx.QueryExecModeSimpleProtocol to be added as pgx will otherwise wrap
+	// the query as a prepared statement, which does *not* support running more than a single statement at a time.
+	return srqf.rewriteError(srqf.wrapped.ExecFunc(ctx, tagFunc, srqf.addAssertToSQL(sql), append([]interface{}{pgx.QueryExecModeSimpleProtocol}, args...)...))
+}
+
+func (srqf strictReaderQueryFuncs) QueryFunc(ctx context.Context, rowsFunc func(ctx context.Context, rows pgx.Rows) error, sql string, args ...any) error {
+	return srqf.rewriteError(srqf.wrapped.QueryFunc(ctx, rowsFunc, srqf.addAssertToSQL(sql), append([]interface{}{pgx.QueryExecModeSimpleProtocol}, args...)...))
+}
+
+func (srqf strictReaderQueryFuncs) QueryRowFunc(ctx context.Context, rowFunc func(ctx context.Context, row pgx.Row) error, sql string, args ...any) error {
+	return srqf.rewriteError(srqf.wrapped.QueryRowFunc(ctx, rowFunc, srqf.addAssertToSQL(sql), append([]interface{}{pgx.QueryExecModeSimpleProtocol}, args...)...))
+}
+
+func (srqf strictReaderQueryFuncs) rewriteError(err error) error {
+	if err == nil {
+		return nil
+	}
+
+	var pgerr *pgconn.PgError
+	if errors.As(err, &pgerr) {
+		if (pgerr.Code == pgInvalidArgument && strings.Contains(pgerr.Message, "is in the future")) ||
```

Comment:
Here the [PG codebase returning](https://github.com/postgres/postgres/blob/1bf29f51fa1a15bd5d28e64070e6e8d105338fc3/src/backend/utils/adt/xid8funcs.c#L121-L125) `is in the future`. It sucks we don't have a more specific error code, but at least we are returning the error anyway and failing closed.

I'd suggest adding unit tests to assert the behaviour in this method. I understand it's covered by `StrictReadModeTest`, but only partially. Does not cover:
- `"replica missing revision"`
- unexpected error

Perhaps we could create a test that just talks to postgres, creates a new transaction that is left opened, and we assert that we get the `replica missing revision` error because of an in-progress transaction.

I also am not sure `"replica missing revision"` is the right message here since this condition could also happen in a primary.

Reply:
Unexpected error would just fallthrough; I can add a test for that if you like (like trying to find a namespace that doesn't exist).

The missing revision will be very hard to test

Reply:
Yes please add error for fallthrough error.

I understand it's difficult to test, but also how do we know it works?

Reply:
I manually tested it, but I'll add another test